File: download/src/main/java/com/example/download/db/ThreadDAOImpl.java
package com.example.download.db;

import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import com.example.download.bean.ThreadInfo;

import java.util.ArrayList;
import java.util.List;

/**
 * @author ShenBF
 * @描述: 数据访问接口实现
 * @date 2018/4/21
 */
public class ThreadDAOImpl implements ThreadDAO {

    private DBHelper helper;

    public ThreadDAOImpl(Context context) {
        helper = new DBHelper(context);
    }

    @Override
    public void insertThread(ThreadInfo threadInfo) {
        SQLiteDatabase db = helper.getWritableDatabase();
        db.execSQL("insert into thread_info(thread_id, url, start, end, finished) values(?, ?, ?, ?, ?)",
                new Object[]{threadInfo.id, threadInfo.url, threadInfo.start, threadInfo.end, threadInfo.finished});
        db.close();
    }

    @Override
    public void deleteThread(String url, int thread_id) {
        SQLiteDatabase db = helper.getWritableDatabase();
        db.execSQL("delete from thread_info where url = ? and thread_id = ?",
                new Object[]{url, thread_id});
        db.close();
    }

    @Override
    public void updateThread(String url, int thread_id, int finished) {
        SQLiteDatabase db = helper.getWritableDatabase();
        db.execSQL("update thread_info set finished = ? where url = ? and thread_id = ?",
                new Object[]{finished, url, thread_id});
        db.close();
    }

    @Override
    public List<ThreadInfo> getThreads(String url) {
        SQLiteDatabase db = helper.getWritableDatabase();
        List<ThreadInfo> list = new ArrayList<>();
        Cursor cursor = db.rawQuery("select * from thread_info where url = ?",
                new String[]{url});
        while (cursor.moveToNext()) {
            ThreadInfo threadInfo = new ThreadInfo(
                    cursor.getInt(cursor.getColumnIndex("thread_id")),
                    cursor.getString(cursor.getColumnIndex("url")),
                    cursor.getInt(cursor.getColumnIndex("start")),
                    cursor.getInt(cursor.getColumnIndex("end")),
                    cursor.getInt(cursor.getColumnIndex("finished"))
            );
            list.add(threadInfo);
        }
        cursor.close();
        db.close();
        return list;
    }

    @Override
    public boolean isExists(String url, int thread_id) {
        SQLiteDatabase db = helper.getWritableDatabase();
        Cursor cursor = db.rawQuery("select * from thread_info where url = ? and thread_id = ?",
                new String[]{url, String.valueOf(thread_id)});
        boolean exists = cursor.moveToNext();
        cursor.close();
        db.close();
        return exists;
    }

}

File: library/download/src/main/java/com/example/download/db/ThreadDAOImpl.kt
package com.example.download.db

import android.content.Context
import com.example.download.bean.ThreadInfo
import java.util.*

/**
 * @author fqxyi
 * @date 2018/4/21
 * 数据访问接口实现
 */
class ThreadDAOImpl(context: Context?) : ThreadDAO {

    private val helper: DBHelper = DBHelper(context)

    override fun insertThread(threadInfo: ThreadInfo) {
        val db = helper.writableDatabase
        db.execSQL(
            "insert into thread_info(thread_id, url, start, end, finished) values(?, ?, ?, ?, ?)",
            arrayOf(
                threadInfo.id,
                threadInfo.url,
                threadInfo.start,
                threadInfo.end,
                threadInfo.finished
            )
        )
        db.close()
    }

    override fun deleteThread(url: String?, thread_id: Int) {
        val db = helper.writableDatabase
        db.execSQL(
            "delete from thread_info where url = ? and thread_id = ?",
            arrayOf(url, thread_id)
        )
        db.close()
    }

    override fun updateThread(
        url: String?,
        thread_id: Int,
        finished: Int
    ) {
        val db = helper.writableDatabase
        db.execSQL(
            "update thread_info set finished = ? where url = ? and thread_id = ?",
            arrayOf(finished, url, thread_id)
        )
        db.close()
    }

    override fun getThreads(url: String?): List<ThreadInfo?>? {
        val db = helper.writableDatabase
        val list: MutableList<ThreadInfo?> =
            ArrayList()
        val cursor =
            db.rawQuery("select * from thread_info where url = ?", arrayOf(url))
        while (cursor.moveToNext()) {
            val threadInfo =
                ThreadInfo(
                    cursor.getInt(cursor.getColumnIndex("thread_id")),
                    cursor.getString(cursor.getColumnIndex("url")),
                    cursor.getInt(cursor.getColumnIndex("start")),
                    cursor.getInt(cursor.getColumnIndex("end")),
                    cursor.getInt(cursor.getColumnIndex("finished"))
                )
            list.add(threadInfo)
        }
        cursor.close()
        db.close()
        return list
    }

    override fun isExists(url: String?, thread_id: Int): Boolean {
        val db = helper.writableDatabase
        val cursor = db.rawQuery(
            "select * from thread_info where url = ? and thread_id = ?",
            arrayOf(url, thread_id.toString())
        )
        val exists = cursor.moveToNext()
        cursor.close()
        db.close()
        return exists
    }

}